server/datastore/mysql/migrations/tables/20230315163954_SoftwareCPEUnqConstraint.go

```diff
@@ -0,0 +1,84 @@
+package tables
+
+import (
+	"database/sql"
+
+	"github.com/fleetdm/fleet/v4/server"
+	"github.com/pkg/errors"
+)
+
+func init() {
+	MigrationClient.AddMigration(Up_20230315163954, Down_20230315163954)
+}
+
+func Up_20230315163954(tx *sql.Tx) error {
+	// Since we will be adding a uniqueness constrain on (software_id) on the software_cpe table - we need to remove any
+	// possible duplicates. Also because there's a chance we remove the duplicate rows before adding
+	// the constraint and new duplicates get generated in between, we need to try to acquire the
+	// vulnerability lock. In case the lock can't be acquired a warning is issued and the migration
+	// will proceed without it.
+
+	removeDuplicatedSoftwareCPEs := func(tx *sql.Tx) error {
+		const deleteStmt = `
+DELETE sc
+FROM software_cpe sc
+	INNER JOIN (
+		SELECT
+			software_id,
+			MAX(id) as max_id
+		FROM software_cpe
+		GROUP BY software_id
+		HAVING COUNT(*) > 1
+	) sc2 ON sc2.software_id = sc.software_id
+WHERE sc.id < sc2.max_id;`
```

Comment:
Curious as to why is `WHERE sc.id < sc2.max_id` needed.

Reply:
Good catch - yeah because of the `HAVING COUNT(*) > 1` that where is not needed.

Reply:
Done in https://github.com/fleetdm/fleet/pull/10514/commits/a094850c9104e2513b63fb04d5a95ff1ff2c4454

Reply:
The WHERE clause is needed. Since we are joining by the `software_id` col, in case of duplicates we will end up with something like:

| id  | software_id   |
|---|---|
|  1 | 1  |
|  2 | 1  |
|  3 | 1  |

Because of the were clause all rows with id less than 3 will be removed.